source/distributed/grid_tools.cc

```diff
@@ -0,0 +1,107 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/point.h>
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/mpi.h>
+#include <deal.II/distributed/grid_tools.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+#ifdef DEAL_II_WITH_MPI
+
+namespace parallel
+{
+  namespace GridTools
+  {
+    template<int spacedim>
+    std::vector< std::vector< BoundingBox<spacedim> > >
+    exchange_local_bounding_boxes(const std::vector< BoundingBox<spacedim> > &local_bboxes,
+                                  MPI_Comm                              mpi_communicator)
+    {
+      // Step 1: preparing data to be sent
+      unsigned int n_bboxes = local_bboxes.size();
+      // data array stores each entry of each point describing the bounding boxes
+      std::vector<double> loc_data_array;
+      loc_data_array.resize(2*spacedim*n_bboxes);
+      for (unsigned int i=0; i<n_bboxes; ++i)
+        for (unsigned int d=0; d < spacedim; ++d)
+          {
+            // Extracting the coordinates of each boundary point
+            loc_data_array[2*i*spacedim + d] = local_bboxes[i].get_boundary_points().first[d];
+            loc_data_array[2*i*spacedim + spacedim + d] = local_bboxes[i].get_boundary_points().second[d];
+          }
+
+      // Step 2: exchanging the size of local data
+      unsigned int n_procs = Utilities::MPI::n_mpi_processes(mpi_communicator);
+
+      // Dimension of the array to be exchanged (number of double)
+      int n_local_data = loc_data_array.size();
```

Comment:
You could move this to line 36, set it right away to `n_local_data = 2*spacedim*n_bboxes`, and then use the value to initialize `loc_data_array`. You don't have to do it.